crates/astria-sequencer-relayer/src/relayer/write.rs

```diff
@@ -0,0 +1,366 @@
+    let mut blobs = Vec::new();
+    block
+        .try_to_blobs(&mut blobs)
+        .wrap_err("failed converting sequencer block to celestia blobs")?;
+    Ok(blobs)
+}
+
+#[instrument(
+    skip_all,
+    fields(
+        num_blobs = blobs.len(),
+        sequencer_heights = %ReportSequencerHeights(&sequencer_heights),
+))]
+async fn submit_blobs(
+    client: HttpClient,
+    blobs: Vec<Blob>,
+    sequencer_heights: Vec<SequencerHeight>,
+    state: Arc<super::State>,
+) -> eyre::Result<u64> {
+    let height = match submit_with_retry(client, blobs, state.clone()).await {
+        Err(error) => {
+            let message = "failed submitting blobs to Celestia";
+            error!(%error, message);
+            return Err(error.wrap_err(message));
+        }
+        Ok(height) => height,
+    };
+    state.set_celestia_connected(true);
+    state.set_latest_confirmed_celestia_height(height);
+    info!(celestia_height = %height, "successfully submitted blocks to Celestia");
+    Ok(height)
+}
+
+async fn submit_with_retry(
+    client: HttpClient,
+    blobs: Vec<Blob>,
+    state: Arc<super::State>,
+) -> eyre::Result<u64> {
+    use celestia_client::{
+        celestia_rpc::BlobClient as _,
+        celestia_types::blob::SubmitOptions,
+    };
+    // Moving the span into `on_retry`, because tryhard spawns these in a tokio
+    // task, losing the span.
+    let span = Span::current();
+    let retry_config = tryhard::RetryFutureConfig::new(u32::MAX)
+        .exponential_backoff(Duration::from_millis(100))
+        // 12 seconds is the Celestia block time
+        .max_delay(Duration::from_secs(12))
+        .on_retry(
+            |attempt: u32, next_delay: Option<Duration>, error: &eyre::Report| {
+                let state = Arc::clone(&state);
+                state.set_celestia_connected(false);
+
+                let wait_duration = next_delay
+                    .map(humantime::format_duration)
+                    .map(tracing::field::display);
+
+                metrics::counter!(crate::metrics_init::CELESTIA_SUBMISSION_FAILURE_COUNT)
+                    .increment(1);
+
+                warn!(
+                    parent: &span,
+                    attempt,
+                    wait_duration,
+                    %error,
+                    "failed submitting blobs to Celestia; retrying after backoff",
+                );
+                futures::future::ready(())
+            },
+        );
+
+    let blobs = Arc::new(blobs);
+    let height = tryhard::retry_fn(move || {
+        let client = client.clone();
+        let blobs = blobs.clone();
+        async move {
+            client
+                .blob_submit(
+                    &blobs,
+                    SubmitOptions {
+                        fee: None,
+                        gas_limit: None,
+                    },
+                )
+                .await
+                .wrap_err("failed submitting sequencer blocks to celestia")
```

Comment:
for a future PR, maybe we should check what kind of error is returned, eg a nonce error or other error, and handle based on that

Reply:
We should. Unfortunately Celestia didn't spec that out anywhere which makes this difficult. :(